docs/design/rubygems.md

```diff
@@ -0,0 +1,528 @@
+
+```ruby
+# :default group
+gem 'nokogiri'
+
+group :test do
+  gem 'faker'
+  gem 'rspec'
+end
+```
+
+Another way to declare a dependency in the `:development` group is to
+[add it](https://guides.rubygems.org/specification-reference/#add_development_dependency) to the `Gem::Specification`,
+which is usually declared in the `.gemspec` file. This means we can safely assume that all dependencies under
+`:development` are dev dependencies.
+
+### Dependency checksums
+The support to checksums in the `Gemfile.lock` is still in development, and currently is an
+[opt-in feature](https://github.com/rubygems/rubygems/pull/7217). To enable it, we need to manually add a `CHECKSUMS`
+section in the `Gemfile.lock`:
+
+```shell
+# manually add `CHECKSUMS` section somewhere in the Gemfile.lock
+vim Gemfile.lock
+# install any gem
+bundle add rails --version "6.1.7"
+# check the Gemfile.lock /o\
+cat Gemfile.lock
+```
+
+Example of a checksum section in the `Gemfile.lock`:
+
+```
+CHECKSUMS
+  actioncable (6.1.7) sha256=ee5345e1ac0a9ec24af8d21d46d6e8d85dd76b28b14ab60929c2da3e7d5bfe64
+  actionmailbox (6.1.7) sha256=c4364381e724b39eee3381e6eb3fdc80f121ac9a53dea3fd9ef687a9040b8a08
+  actionmailer (6.1.7) sha256=5561c298a13e6d43eb71098be366f59be51470358e6e6e49ebaaf43502906fa4
+  actionpack (6.1.7) sha256=3a8580e3721757371328906f953b332d5c95bd56a1e4f344b3fee5d55dc1cf37
+  actiontext (6.1.7) sha256=c5d3af4168619923d0ff661207215face3e03f7a04c083b5d347f190f639798e
+  actionview (6.1.7) sha256=c166e890d2933ffbb6eb2a2eac1b54f03890e33b8b7269503af848db88afc8d4
+```
+
+This feature is available since Bundler [v2.5.0](https://github.com/rubygems/rubygems/blob/master/bundler/lib/bundler/lockfile_parser.rb#L55),
+from this [PR](https://github.com/rubygems/rubygems/pull/6374) being merged on Oct 21, 2023.
+
+## II. Overview of the current implementation in Cachito
+
+[cachito/workers/pkg_mangers/rubygems.py](https://github.com/containerbuildsystem/cachito/blob/master/cachito/workers/pkg_managers/rubygems.py)
+
+Most work is already done by parsing the `Gemfile.lock` file, which pins all dependencies to exact versions. The only
+supported source for gem dependencies to be fetched from is <https://rubygems.org>. Git dependencies are specified
+using a repo URL and pinned to a commit hash. Path dependencies are specified using a local path.
+
+To avoid arbitrary code execution, Bundler **is not used** to download dependencies. Instead, as stated above, Cachito
+parses `Gemfile.lock` file directly and download the gems from <https://rubygems.org>.
+
+**Note**: parsing `Gemfile.lock` is done via [gemlock-parser](https://github.com/containerbuildsystem/gemlock-parser),
+which is vendored from
+[scancode-toolkit](https://github.com/nexB/scancode-toolkit/blob/develop/src/packagedcode/gemfile_lock.py).
+
+Source code for "official" Bundler lockfile parsing in Ruby:
+<https://github.com/rubygems/rubygems/blob/master/bundler/lib/bundler/lockfile_parser.rb>
```

Comment:
We're adding this to cachi2 and yes, much of the logic will be ported, but we don't need to mention cachito here, we're trying hard to drop all references to it ultimately to just mentioning in the main README that "origin of this project was cachito, link here..." but I don't think this should be in the design doc since we probably do end up making different decisions and the implementation will also be different in many cases.

Reply:
I don't think this specific section brings too much value besides linking the main reference we had, the implementation done in Cachito. So I'd be fine with droping it, but I still think there's value to link previous work, specially considering that there was much research done back then. 

Why do you think it's bad to reference Cachito? I think it's ok to say we are taking a different path with Cachi2, and the work done before is merely a reference and a starting point.

Reply:
This,

> there was much research done back then

and the fact that we often borrow starting code from Cachito, make it a good idea to at least provide links.

Reply:
The original project is to be sunset in the future with this one being its replacement. I don't see value in putting many references to the old project we don't really want to maintain live connections to, so while I'm okay stating that parts of the code were borrowed, linking at too many places doesn't IMO achieve what we're ultimately trying to do, because the code, yet borrowed, has to be often significantly adapted to this very code base, so providing links to code which will be 75% reworked is IMO misleading and confusing. It should only ever be nothing more than a kudos/credit to where we got inspired and that's it.

Reply:
So I guess we disagree on this point. I do agree that this section as it was did not bring too much value, so I moved the content to a "References" section in the end of the document, and made it shorter to just convey the credits and links to the original project.